### ISSUE TITLE
SimCoupe behaving differently with new Z80 core

### ISSUE THREAD

**Issue by stefandrissen** [simonowen/simcoupe#54]
Sorry for this being a really really lousy test case, but my code (attached) is crashing differently with the 20210624 dev build compared to the 20210501 build.

With the dev build it is running around in high memory page 0x0e, while with the previous build it is running around in low memory page 08 with page 0x10 in high memory.

[lllll.zip](https://github.com/simonowen/simcoupe/files/6759842/lllll.zip)

**Comment by simonowen** [comment]
Thanks for the disk! Anything that even occasionally gives an issue is a good starting point.

I think I can see it happening on the title screen, when the tune freezes on one note. It looks like it's maybe preparing data in the background during that time? I had a step-out active at the time so there was an instruction trace, though the results looked a bit confusing, and I'll continue digging.

The new CPU core should be pretty good overall since it has passed various test suites. Though I did find a bug in OTDR when I first switched, which caused the initial SAM BASIC palette to be wrong. Those are hard to test without devices to read/write from, and I don't think they're covered. I've not seen any other timing differences with timing-sensitive demos, though they probably use a subset of the instruction set.

I was due to release an updated dev build with some debugger fixes but it's definitely worth holding off to include this too.

**Comment by simonowen** [comment]
Peppering a few breakpoints around the start of page 6 I caught a trace that included this:

![image](https://user-images.githubusercontent.com/200089/124390131-21a34580-dce2-11eb-9e56-1e3094692fb2.png)

It shows the normal IM 1 interrupt executing junk at a location that previously held valid handler code.  Putting a memory write breakpoint on the start of the handler when its valid showed this:

![image](https://user-images.githubusercontent.com/200089/124390242-89599080-dce2-11eb-8771-f356bd866e40.png)

That seems consistent with the first trace, showing the code that is overwriting the interrupt handler. Do you recognise where that is in your code?

I'd guess that the execution behaviour will be fairly variable, depending on how much code had been overwritten before the interrupt handler triggered. With any luck if the overwriting is fixed both versions will run fine.

I might still see if I can stabilise the starting conditions to check that there's no other difference between the two versions.

**Comment by stefandrissen** [comment]
In the mean time I've managed to revert my code back to a non-crashing state, but have yet to figure out where I messed up.

As to the above, page 6 is for screen memory, does not contain an interrupt handler and should not manage to get itself paged into section A.

**Comment by simonowen** [comment]
If I set `bpm 6:18 w` and `bpm 6:19` on the boot screen, I get hits to each during loading, but 6:0019 is the first location hit at the problem point in the tune. I get the same behaviour with both releases you're using so I'm not seeing that difference. Is it repeatable?

I can see how writing to 6:0018 would give a difference in paging though!

Just in case it's seen as a difference, bug #53 causes the debugger to step just the index prefix and show the underlying HL instruction separately. That only affects the dev version and is fixed for the next build.

**Comment by simonowen** [comment]
Another test I tried was `bpmr 6:0 1a w` to watch the first 26 bytes of page 6. There is one write to 6:0000, but the first write after that is to 6:0019 again. I've tried 4 different SimCoupe versions and repeated at least 10 times in each with the same result.

Are you definitely seeing a 6:0018 write using the same disk image from above?

**Comment by stefandrissen** [comment]
Right... had another look tonight using the lllll.dsk image above using the 05-29 and the 06-24 builds.

When title screen shown

`bpm 06:19 w` 
continue
`bpx 38` 
step to the `rst 18`

Now both show `jp m,11d1` but only 05-29 considers the `m` flag set and jumps, 06-24 does not consider the `m` flag set.

**Comment by simonowen** [comment]
I've traced the main cause of the difference you described above, such that the registers and flags now match when it hits the `JP M` instruction. The offset to where the main screen contention is applied was off by a few cycles, causing the non-interrupt code alignment to slip relative to the interrupt trigger points. When the interrupt handler was overwritten enough to cause problems the flags from the non-interrupt code was then controlling the `JP M` condition. Any change of interrupt trigger point due to timing differences would mean the register values seen at 6:18 could be wildly different between versions.

Surprisingly my ASIC test program still passes on the dev version, and all the timing-sensitive demos I tried also work without issue, so it's a _very_ subtle issue. Despite the pain, I'm glad it's been found before it's been released more widely. I'm going to write some new tests that poke at the area, making sure it fails on the dev release and works on real hardware and other releases. Ultimately these will become unit tests for the emulator build to protect against similar breakages.

While I was looking at the main issue I noticed an R register difference creeping in somewhere too, so I need to check whether it's the old or new core that's correct. Once I've checked that and have the new tests working I'll build a new release from it all.

The change to fix Anton's demo was the generated sound sample data, which was slightly the wrong frequency, due to ratio rounding in the 3rd party code SimCoupe uses. It affected only what you heard in the combined sound output from SimCoupe and didn't otherwise affect the running code.

None of this directly helps you track down the interrupt handler being overwritten though. I'd trust the pre-dev version for now, which does match the behaviour of my current development build with the offset fix.
